src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -53,33 +56,65 @@ public override void Initialize (AnalysisContext context)
-		{
-			var targetMethod = invocationOperation.TargetMethod;
-			if (targetMethod.HasAttribute (RequiresUnreferencedCodeAnalyzer.RequiresUnreferencedCodeAttribute))
-				return;
+			ImmutableArray<ITypeParameterSymbol> typeParams = default;
+			ImmutableArray<ITypeSymbol> typeArgs = default;
+			if (context.SemanticModel.GetTypeInfo (context.Node).ConvertedType is INamedTypeSymbol type) {
+				// INamedTypeSymbol inside nameof, commonly used in [ExpectedWarning] can generate diagnostics
+				// Walking the node heirarchy to check if INamedTypeSymbol is inside a nameof
+				var node = context.Node;
+				while (node != null) {
+					if (node is InvocationExpressionSyntax invocationExpression && invocationExpression.Expression is IdentifierNameSyntax ident1) {
+						if (ident1.Identifier.ValueText.Equals ("nameof"))
+							return;
+					}
+					node = node.Parent;
+				}
+				typeParams = type.TypeParameters;
+				typeArgs = type.TypeArguments;
+			} else if (context.SemanticModel.GetSymbolInfo (context.Node, context.CancellationToken).Symbol is IMethodSymbol targetMethod) {
+				// Check if the generic parameters have any constraints on this method
+				var node = context.Node;
+				while (node != null) {
+					if (node is MethodDeclarationSyntax parentMethod) {
+						foreach (var constraintCaluse in parentMethod.ConstraintClauses) {
+							foreach (var constraint in constraintCaluse.Constraints) {
+								if (constraint.DescendantTokens ().Where (t => t.Kind () == SyntaxKind.StructKeyword ||
+								t.Kind () == SyntaxKind.NewKeyword).Any ())
+									return;
+								if (constraint is TypeConstraintSyntax typeConstraint &&
+									typeConstraint.Type is IdentifierNameSyntax typeConstraintTdentifier &&
+									typeConstraintTdentifier.Identifier.Text.Equals ("unmanaged"))
+									return;
+							}
+						}
+						// Exit loop if the declared method doesn't have any constraints
+						break;
+					}
+					node = node.Parent;
+				}
+				typeParams = targetMethod.TypeParameters;
+				typeArgs = targetMethod.TypeArguments;
+			}
 
-			for (int i = 0; i < targetMethod.TypeParameters.Length; i++) {
-				var sourceValue = GetTypeValueNodeFromGenericArgument (targetMethod.TypeArguments[i]);
-				var targetValue = new GenericParameterValue (targetMethod.TypeParameters[i]);
-				foreach (var diagnostic in GetDynamicallyAccessedMembersDiagnostics (sourceValue, targetValue, invocationOperation.Syntax.GetLocation ()))
-					context.ReportDiagnostic (diagnostic);
+			if (typeParams != null) {
+				Debug.Assert (typeParams.Length == typeArgs.Length);
+				for (int i = 0; i < typeParams.Length; i++) {
+					var sourceValue = GetTypeValueNodeFromGenericArgument (typeArgs[i]);
+					var targetValue = new GenericParameterValue (typeParams[i]);
+					foreach (var diagnostic in GetDynamicallyAccessedMembersDiagnostics (sourceValue, targetValue, context.Node.GetLocation ()))
```

Comment:
I think maybe we should use the location of the type parameter instead of the whole generic type name - I think the more specific locations are a better UX.

Reply:
As discussed this doesn't look straightforward so I am ok with reporting the full generic name as the location for now. I filed this issue to track the suggestion: https://github.com/dotnet/linker/issues/2569